### ISSUE TITLE
renv::install() doesn't honor type = 'binary' argument and fails to provide warning or error

### ISSUE THREAD

**Issue by johnkarlen** [rstudio/renv#1046]
**expected behavior**: when setting `type = 'binary'` argument for `renv::install()`, I expect it to find a binary version of package, or provide warning or error about what went wrong.

**actual behavior**: always downloads source version of specified package version with no warning, error, or explanation.

My coworker is able to retrieve a binary version of the package when running this command on his machine, but mine only finds a source version with no explanation.

```
> renv::install('stringi@1.6.2', rebuild = T, type = 'binary', repos = "https://cran.microsoft.com")
* Querying repositories for available source packages ... Done!
Retrieving 'https://cran.microsoft.com/src/contrib/Archive/stringi/stringi_1.6.2.tar.gz' ...
        OK [file is up to date]
Installing stringi [1.6.2] ...
```

If I edit the renv.lock file and specify the url that my coworker is able to download the binary version from, I am able to install the binary. This also works for me when I install from the url directly.`install.packages(URL, type = 'binary', repos = NULL)`: 

```
    "stringi": {
      "Package": "stringi",
      "Version": "1.6.2",
      "Source": "URL",
      "RemoteUrl": "https://cran.microsoft.com/snapshot/2021-06-29/bin/macosx/contrib/4.1/stringi_1.6.2.tgz",
      "Repository": "CRAN",
      "Hash": "9df5e6f9a7fa11b84adf0429961de66a",
      "Requirements": []
    },

```

I'm running:
Mac OS  12.3.1 (Intel)
R version 4.2.1
Rstudio 2022.02.3 

I get similar results when running on the rocker r-base image. 

```
> renv::install('stringi@1.6.2', type = 'binary')
Retrieving 'https://cran.rstudio.com/src/contrib/Archive/stringi/stringi_1.6.2.tar.gz' ...
        OK [downloaded 7.2 Mb in 1.1 secs]
...
```

```
> renv::install('stringi@1.6.2', type = 'binary', repos = "https://cran.microsoft.com")
Warning: failed to find binary for 'stringi 1.6.2' in package repositories
Warning: failed to find source for 'stringi 1.6.2' in package repositories
Error: failed to retrieve package 'stringi@1.6.2'
Traceback (most recent calls last):
7: renv::install("stringi@1.6.2", type = "binary", repos = "https://cran.microsoft.com")
6: retrieve(names(remotes))
5: handler(package, renv_retrieve_impl(package))
4: renv_retrieve_impl(package)
3: renv_retrieve_repos(record)
2: stopf("failed to retrieve package '%s'", renv_record_format_remote(record))
1: stop(sprintf(fmt, ...), call. = call.)
```

**Comment by shearerpmm** [comment]
If you're not able to reproduce this issue with stringi@1.6.2, maybe use a package/version combination that does NOT have a binary on in the target repository. When we specifically ask for a binary, we would prefer to get a warning or error (or at least an option to get a warning/error) when renv is unable to retrieve a binary, rather than silently falling back on source. This would help us configure and debug our deployment process more robustly, since source packages can be much harder to deploy than binaries.

**Comment by kevinushey** [comment]
There are a couple things going on here...

You're using R 4.2.1, but apparently trying to install binaries for R 4.1. This is a bad idea; binaries for older versions of R are not guaranteed to work with newer versions of R.

The fact that `renv` is trying to install the package from the archive is a bug in `renv` though; I'll take a look.

**Comment by shearerpmm** [comment]
Good point - John also ran this in R 4.1 using rocker and had the same issue.

**Comment by johnkarlen** [comment]
```~ % docker run --rm -ti rocker/r-ver:4.1.3

R version 4.1.3 (2022-03-10) -- "One Push-Up"
Copyright (C) 2022 The R Foundation for Statistical Computing
Platform: aarch64-unknown-linux-gnu (64-bit)

...

> install.packages('renv')
Installing package into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://packagemanager.rstudio.com/cran/2022-04-21/src/contrib/renv_0.15.4.tar.gz'
Content type 'binary/octet-stream' length 969163 bytes (946 KB)
==================================================
downloaded 946 KB

* installing *source* package ‘renv’ ...
** package ‘renv’ successfully unpacked and MD5 sums checked
** using staged installation
** R
** inst
** byte-compile and prepare package for lazy loading
** help
*** installing help indices
*** copying figures
** building package indices
** installing vignettes
** testing if installed package can be loaded from temporary location
** testing if installed package can be loaded from final location
** testing if installed package keeps a record of temporary installation path
* DONE (renv)

The downloaded source packages are in
	‘/tmp/RtmpxgfGej/downloaded_packages’
> renv::install('stringi@1.6.2', type = 'binary')
* Querying repositories for available binary packages ... Done!
* Querying repositories for available source packages ... Done!
Retrieving 'https://packagemanager.rstudio.com/cran/__linux__/focal/2022-04-21/src/contrib/Archive/stringi/stringi_1.6.2.tar.gz' ...
	OK [downloaded 2.8 Mb in 1.6 secs]
Installing stringi [1.6.2] ...
	OK [installed binary]
Copying stringi [1.6.2] into the cache ...
	OK [copied to cache in 7.8 milliseconds]
>```

**Comment by kevinushey** [comment]
In that case, R is installing packages from the archive using RSPM binaries.

RSPM works by "disguising" binary packages as source packages; even though the repository URL is "technically" a source repository, the packages being served are binary packages.

```
> renv::install('stringi@1.6.2', type = 'binary')
* Querying repositories for available binary packages ... Done!
* Querying repositories for available source packages ... Done!
Retrieving 'https://packagemanager.rstudio.com/cran/__linux__/focal/2022-04-21/src/contrib/Archive/stringi/stringi_1.6.2.tar.gz' ...
	OK [downloaded 2.8 Mb in 1.6 secs]
Installing stringi [1.6.2] ...
	OK [installed binary]
Copying stringi [1.6.2] into the cache ...
	OK [copied to cache in 7.8 milliseconds]
```

I think there's a bunch of things to keep in mind here.

- CRAN only builds and publishes binaries for Windows + macOS.
- MRAN is a service that snapshots CRAN, and so Windows + macOS binaries associated with older CRAN snapshots can be installed.
- RSPM is an RStudio service that provides binaries on most platforms, including Linux. It does so through a separate mechanism, where binary packages are disguised as source packages.

In other words, as far as R is concerned on Linux, "binary" repositories aren't a thing -- only "source" repositories exist. RSPM works around this by using custom repository URLs where platform-specific package binaries are served.

**Comment by johnkarlen** [comment]
yes, also seeing it pull source on my M1 mac running R 4.1.2

```The downloaded source packages are in
	‘/private/var/folders/b2/_v51_7m14g9fl8tvj05c9dlr0000gn/T/RtmpR0yqHy/downloaded_packages’
> renv::install('stringi@1.6.2', type = 'binary')
* Querying repositories for available binary packages ... Done!
* Querying repositories for available source packages ... Done!
Retrieving 'https://cran.case.edu/src/contrib/Archive/stringi/stringi_1.6.2.tar.gz' ...
	OK [downloaded 7.2 Mb in 1.2 secs]
Installing stringi [1.6.2] ...
```

**Comment by shearerpmm** [comment]
@johnkarlen and that behavior is not fixed by [a55c138](https://github.com/rstudio/renv/commit/a55c1384ad1ae367d2fdce3ba0b409c8cac0790d)?

**Comment by johnkarlen** [comment]
I get the expected behavior from [a55c138](https://github.com/rstudio/renv/commit/a55c1384ad1ae367d2fdce3ba0b409c8cac0790d). It cannot find a binary on my M1 mac running 4.1.2.
```
* Querying repositories for available binary packages ... Done!
Warning: failed to find binary for 'stringi 1.6.2' in package repositories
Error: failed to retrieve package 'stringi@1.6.2'
Traceback (most recent calls last):
7: renv::install("stringi@1.6.2", type = "binary")
6: records <- retrieve(names(remotes)) at install.R#207
5: for (package in packages)
     handler(package, renv_retrieve_impl(package)) at retrieve.R#33
4: handler = handler %||% function(package, action) action, at restore.R#267
3: switch(source,
          bioconductor = renv_retrieve_bioconductor(record),
          bitbucket    = renv_retrieve_bitbucket(record),
          git          = renv_retrieve_git(record),
          github       = renv_retrieve_github(record),
          gitlab       = renv_retrieve_gitlab(record),
          repository   = renv_retrieve_repos(record),
          url          = renv_retrieve_url(record),
          renv_retrieve_unknown_source(record)
   ) at retrieve.R#184
2: stopf("failed to retrieve package '%s'", renv_record_format_remote(record)) at retrieve.R#603
1: stop(sprintf(fmt, ...), call. = call.) at utils-format.R#13
```

**Comment by shearerpmm** [comment]
I think that's because stringi@1.6.2 doesn't have an ARM binary anywhere. This issue can probably be closed, thanks so much @kevinushey!

**Comment by kevinushey** [comment]
Thanks!
